src/roi/Roi.ts

```diff
@@ -91,4 +138,584 @@ export class Roi {
+  }
+  /**
+   * Getter that calculates sphericity of the ROI
+   */
+  get sphericity() {
+    return (2 * Math.sqrt(this.surface * Math.PI)) / this.perimeter;
+  }
+
+  /**
+   * Getter that calculates solidity of the ROI
+   */
+  get solidity() {
+    return this.surface / getConvexHull(this.getMask()).surface;
+  }
+
+  get convexHull() {
+    return this.#getComputed('convexHull', () => {
+      return getConvexHull(this.getMask());
+    });
+    // if (!this.#computed.convexHull) {
+    //   this.#computed.convexHull = getConvexHull(this.getMask());
+    // }
+    // return this.#computed.convexHull;
+  }
+
+  get mbr() {
+    return this.#getComputed('mbr', () => {
+      return getMbr(this.getMask());
+    });
+    // if (!this.#computed.mbr) {s
+    //   this.#computed.mbr = getMbr(this.getMask());
+    // }
+    // return this.#computed.mbr;
+  }
+
+  get roundness() {
+    /*Slide 24 https://static.horiba.com/fileadmin/Horiba/Products/Scientific/Particle_Characterization/Webinars/Slides/TE011.pdf */
+    return (4 * this.surface) / (Math.PI * this.feret.maxDiameter.length ** 2);
+  }
+
+  get feret(): Feret {
+    return this.#getComputed('feret', () => {
+      return getFeret(this.getMask());
+    });
+    // if (!this.#computed.feret) {
+    //   this.#computed.feret = getFeret(this.getMask());
+    // }
+    // return this.#computed.feret;
+  }
+  /**
+   *
+   * @returns calculated properties as one object
+   */
+  toJSON() {
+    return {
+      id: this.id,
+      height: this.height,
+      width: this.width,
+      surface: this.surface,
+      eqpc: this.eqpc,
+      ped: this.ped,
+      feret: this.feret,
+      fillRatio: this.fillRatio,
+      sphericity: this.sphericity,
+      roundness: this.roundness,
+      solidity: this.solidity,
+      perimeter: this.perimeter,
+      convexHull: this.convexHull,
+      mbr: this.mbr,
+    };
+  }
+
+  #getComputed<T extends keyof Computed>(
+    property: T,
+    callback: () => Computed[T],
+  ): Computed[T] {
+    if (this.#computed[property] === undefined) {
+      let result = callback();
+      this.#computed[property] = result;
+      return result;
+    }
+    return this.#computed[property] as Computed[T];
+  }
+}
+
+/**
+ *
+ * @param value
+ * @param message
+ */
+export function assert(value: unknown, message?: string): asserts value {
+  if (!value) {
+    throw new Error(`unreachable${message ? `: ${message}` : ''}`);
+  }
+}
+
+/**
+ *
+ * @param roi -ROI
+ * @returns object which tells how many pixels are exposed externally to how many sides
+ */
+function getPerimeterInfo(roi: Roi) {
+  let roiMap = roi.getMap();
+  let data = roiMap.data;
+  let one = 0;
+  let two = 0;
+  let three = 0;
+  let four = 0;
+
+  for (let column = 0; column < roi.origin.column + roi.width; column++) {
+    for (let row = 0; row < roi.origin.row + roi.height; row++) {
+      let target = computeIndex(roi, row, column);
```

Comment:
I don't think those loop end values are correct if you are using computeIndex.

Reply:
I fixed this issue and allowed. In my screw project it allows to calculate all the paramters in 2.12s instead of 5.47s. I'm checking now what is the next limiting speed step.